android-components/components/feature/addons/src/main/java/mozilla/components/feature/addons/ui/AddonFilePicker.kt

```diff
@@ -0,0 +1,83 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.addons.ui
+
+import android.content.ActivityNotFoundException
+import android.content.Context
+import android.content.Intent
+import android.net.Uri
+import androidx.activity.result.ActivityResultCaller
+import androidx.activity.result.ActivityResultLauncher
+import androidx.activity.result.contract.ActivityResultContracts
+import mozilla.components.feature.addons.Addon
+import mozilla.components.feature.addons.AddonManager
+import mozilla.components.support.base.log.logger.Logger
+import mozilla.components.support.ktx.android.net.toFileUri
+import java.util.UUID
+
+/**
+ * Allows to launch a file picker to select, an add-on file to be installed.
+ * @param context the application context
+ *
+ */
+class AddonFilePicker(
+    val context: Context,
+    private val addonManager: AddonManager,
+) {
+    internal lateinit var activityLauncher: ActivityResultLauncher<Array<String>>
+    private val logger = Logger("AddonFilePicker")
+
+    /**
+     * Launch an Android file picker, where the user can select an add-on file.
+     * @returns a [Boolean] indicating if the file picker was launched successfully or not.
+     */
+    fun launch(): Boolean {
+        return try {
+            activityLauncher.launch(emptyArray())
+            true
+        } catch (e: ActivityNotFoundException) {
+            logger.error("Unable to find an app to select an XPI file", e)
+            false
+        }
+    }
+
+    /**
+     * Registers a lister for file picker results.
+     * @param resultCaller The [ActivityResultCaller] on which results will be observed.
+     */
+    fun registerForResults(resultCaller: ActivityResultCaller) {
+        activityLauncher =
+            resultCaller.registerForActivityResult(AddonOpenDocument(), ::handleUriSelected)
+    }
+
+    internal fun handleUriSelected(uri: Uri?) {
+        uri?.let {
+            val fileUri = convertToFileUri(uri)
+            val onSuccess: ((Addon) -> Unit) = {
+                logger.info("Add-on installed successfully")
+            }
+            val onError: ((String, Throwable) -> Unit) = { _, _ ->
+                logger.info("Unable to install add-on")
+            }
+            addonManager.installAddon(
+                Addon(id = UUID.randomUUID().toString(), downloadUrl = fileUri),
```

Comment:
@willdurand as the `id` is not used as part of the installation process, I added `randomUUID`, after the extension is install the add-on manger provide us correct id from the manifest.

Reply:
Any thought on changing `installAddon()` to accept a `url` param instead of a full (and yet partial) `Addon` instance?

Reply:
I added a new commit with the suggested changes https://github.com/mozilla-mobile/firefox-android/pull/4568/commits/ade1f097bda2d282520bcd7506ee4e68dfa1b10a :)